WIP: Prf bulk entry

tacaswell opened the pull request:
more violence to come

tacaswell committed:
PRF: add bulk insert

tacaswell committed:
PEP8: whitespace

tacaswell committed:
WIP: rework document dereference and caching

tacaswell committed:
WIP: first pass at fast event retrival

tacaswell committed:
MNT: clean up bulk insertion

 - whitespace
 - use new caching of event descriptors

tacaswell committed:
MNT: remove removed function from API

tacaswell committed:
WIP: fix up glaing errors

tacaswell committed:
MNT: remove time_as_datetime from db entry

tacaswell committed:
API: make uid obligatory

tacaswell committed:
API: require new representation of events

@danielballan

tacaswell committed:
TST: remove test of removed function

tacaswell committed:
TST: fix up tests due to API changes

account for uid being required

tacaswell committed:
API: last optional uid

tacaswell committed:
API: remove BeamlineConfig as seperate collection

tacaswell committed:
TST: update travis to include boltons

tacaswell committed:
API: use new caching in all find functions

 - this breaks some bits of the API, these should probably be simplified
   anyway though.

tacaswell committed:
API/MNT: remove old version of Document

tacaswell committed:
MNT: don't look up runstart on insert if cached

tacaswell committed:
PRF: more gofast code

from @arkilic

tacaswell commented on metadatastore/commands.py:
```diff
@@ -345,34 +468,69 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
+
+    """
+
+    descriptor = event_desc_given_uid(event_descriptor)
```
you mean to use `XXX._collection.find_one(...)` in the `_cache_*` functions?

tacaswell replied:
~~that made it slower~~ no difference

tacaswell committed:
MNT: use more stable internal function

`_collection` is part of ME's cache mechanisms, `_get_collection()`
ensures that `_collection` exists and is not `None`.

tacaswell committed:
PRF: sidestep yet more of mongoengine

tacaswell committed:
MNT: restore html repr

tacaswell committed:
MNT: make sure keys in html repr are sorted

tacaswell committed:
MNT: also include ISO time in timestamp

tacaswell committed:
TST: add test for bulk insert + simple generator

tacaswell committed:
MNT: bugs found via dataportal tests

ericdill commented on metadatastore/commands.py:
```diff
@@ -1,31 +1,51 @@
+           'reorganize_event']
+
+# process local caches of 'header' documents these are storing object indexed
+# on Objected because that is what the reference fields in mongo are
```
Is `Objected` a typo?

tacaswell replied:
missing an `I`

ericdill commented on metadatastore/commands.py:
```diff
@@ -169,11 +326,9 @@ def insert_run_start(time, scan_id, beamline_id, beamline_config, uid=None,
 
-    beamline_config = _get_mongo_document(beamline_config, BeamlineConfig)
     run_start = RunStart(time=time, scan_id=scan_id,
-                         time_as_datetime=_todatetime(time), uid=uid,
```
Why are you getting rid of `time_as_datetime`?

tacaswell replied:
Because this is something that should be taken care of on the way out (see the companion dataportal PR) and is already gone from the ODM templates.

This also makes @arkilic 's life easier by keeping the datatypes stored in the db simple.

ericdill replied:
However, having datetimes in the non-event documents makes your life so so so much better whenever you need to drop to a mongo shell and debug

ericdill commented on metadatastore/commands.py:
```diff
@@ -591,7 +703,7 @@ def find_run_starts(**kwargs):
     ----
     All documents that the RunStart Document points to are dereferenced.
-    These include RunStop, BeamlineConfig, and Sample.
+    These include RunStop, and Sample.
```
`Sample` is not a document

ericdill replied:
Also `RunStart` does **not** point `RunStop`

ericdill commented on metadatastore/commands.py:
```diff
@@ -791,24 +858,27 @@ def find_events(descriptor=None, **kwargs):
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
+                                        for j in range(2)]
+        # wrap it our fancy dict
```
-> `wrap it in our fancy dict`

ericdill commented on metadatastore/odm_templates.py:
```diff
@@ -10,29 +10,13 @@
-    an experiment's run. BeamlineConfig is NOT required to create a RunStart
+    an experiment's run.
     The only prereq is an EventDescriptor that identifies the nature of
     event that is starting and
```
Any idea what the rest of this is supposed to say?

tacaswell committed:
MNT: fix up Event indexes

tacaswell committed:
API/PRF: reverse order events are yielded in

For reasons I do not understand, returning the events sorted ascending
by time is fast, descending by time is slow (probably scales with DB
size).

tacaswell commented on metadatastore/doc.py:
```diff
@@ -0,0 +1,79 @@
+    def __contains__(self, k):
+        if isinstance(k, six.string_types) and k.startswith('_'):
+            return False
+        return super(Document, self).__contains(k)
```
indeed, good catch!

tacaswell committed:
FIX: fix typo

tacaswell committed:
MNT: simplify additional indexes

 - uid is implicit because it is unique
 - _id is always indexed by mongo

tacaswell committed:
FIX: len did not respect hidden keys

tacaswell committed:
TST: full test coverage on doc

tacaswell committed:
FIX: deal with RunStart or uid as input

Deal with either a RunStart document (which means we need to grab the
uid) or a uid is input to `find_event_descriptors`.

tacaswell committed:
API: remove _normalize_object_id

We should never get ObjectId objects through the public facing
functions.

tacaswell committed:
MNT: deal with RunStart documents as input

tacaswell committed:
MNT: deal with EventDescriptor documents as input

In `find_events` properly deal with EventDescriptor documents
as input as well as uid strings

tacaswell committed:
FIX: fix ObjectId lookup path

This bug was being masked by the uid-based lookup from the inserts
populating the cache.

tacaswell committed:
PRF: cache RunS*, EventDescriptor on insert

Save a lookup or two

tacaswell committed:
ENH/TST: add clear_process_cache + tests

Clear the cache, tests make sure we hit oid path
Commit 7f31c8b0fd878bb2d751cc639913701ec8d66335 fixed in tests

tacaswell committed:
API: remove _runstop_given_oid

Nothing holds a reference to a runstop document, this should
never be used so removing

tacaswell committed:
MNT: house keeping

 - remove unused commits
 - add runstop to bulk insert test

tacaswell committed:
API: remove unneeded functions

These are public functions to smooth over uid vs oid access, however
nothing outside of this library should know about ObjectIds so there is
no need to have an API to accept them from the outside world.

tacaswell committed:
MNT: factored out logic for doc or uid lookup

tacaswell committed:
ENH: add fetch_events_table

Return data from an event stream as a collection of
DataFrames and Series

tacaswell committed:
API: remove unused function

This is now taken care of directly in `fetch_events_generator`
and `find_events`

tacaswell committed:
MNT: simplify formatting embedded docs for ME

Only use this one place, just do it in-line

tacaswell committed:
MNT/API: forbid name key with '.'

Rather than doing a in/out conversion every time, just forbid there
to be dots in the data key names.

tacaswell committed:
MNT: remove unused import

tacaswell committed:
ENH: {runstop,eventdescriptor}_by_runstart

Implement single-purpose functions to get RunStop and EventDescriptor
documents given a RunStart

tacaswell committed:
FIX: fix bug in find_runstart

Added a test

tacaswell committed:
TST: add simple test for find_run_stops

tacaswell committed:
API: make _doc_or_uid public

rename `_doc_or_uid` -> `doc_or_uid_to_uid`

tacaswell committed:
API: runstop_by_runstart raise custom exception

Raise a NoRunStop type exception

tacaswell committed:
API: make _pretty_print_time public

ericdill commented on metadatastore/commands.py:
```diff
@@ -214,56 +438,27 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
+    runstart_oid = _RUNSTART_UID_to_OID_MAP[run_start['uid']]
+    rs_ref = DBRef('RunStart', runstart_oid)
 
+    run_stop = RunStop(run_start=rs_ref, reason=reason, time=time,
```
while we (you) are tinkering with the mds api, what are your feelings on raising if a second insert_run_stop call gets made for the same run_start?

I (amusingly) put the "insert_run_stop" call nested in the event creation loop for one of the dataportal sample scripts
https://gist.github.com/anonymous/d473d5ab69e3d821f9b4 

...and got an integrity error for 21 run stop records :grinning:

tacaswell committed:
ENH: second RunStop for RunStart raises

ht @ericdill

tacaswell committed:
DOC: typo

single letter diff!

ericdill commented on metadatastore/commands.py:
```diff
@@ -1,31 +1,81 @@
+
+
+def doc_or_uid_to_uid(doc_or_uid):
+    """Helper function to ensure a uid
```
to ensure a uid does what?

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
-    return connect(db=database, host=host, port=port, alias=ALIAS)
 
+def runstop_by_runstart(run_start):
+    """Given a RunStart return a list of it's RunStop
```
Why would this return a list? There better be only one!

ericdill replied:
![496467356_128977421765_xlarge](https://cloud.githubusercontent.com/assets/6223243/9649911/79bfdbf2-51c9-11e5-9497-1f89fd336886.jpeg)

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
+    Parameters
+    ----------
+    descriptor : dict or uid
+        The EeventDestriptor to get the events for.  Can be either
```
typo here --> `EeventDestriptor`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
+    timestamps_table = pd.DataFrame({k: pd.to_datetime(ts[k], unit='s')
+                                     for k in ts})
+    # return the whole lot
+    return descriptor, data_table, times, uids, timestamps_table
```
I'm sure that @danielballan will love that this function returns 5 things :)

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
+
+    uids : Series
+        The uids of each event as a `Series` indexed on sequence number
+
```
drop the extra whitespace between these lines

tacaswell committed:
FIX: transmute KeyError -> AttributeError

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
+    data_table = pd.DataFrame([ev['data'] for ev in all_events], index=seq_num)
+
+    # get timestamps
+    ts = pd.DataFrame([ev['timestamps'] for ev in all_events], index=seq_num)
```
Iterating over `all_events` five times feels a little dirty. Wouldn't it be faster to iterate over it once and build these five lists in the same for loop? Maybe use a deque?

tacaswell replied:
probably

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,88 +89,273 @@ def inner(*args, **kwargs):
+        {'run_start_id': oid})
+
+    # loop over the found documents, cache, and dereference
+    rets = [_cache_eventdescriptor(ev_desc) for ev_desc in ev_desc_cur]
```
Don't you want to check the cache first?

tacaswell replied:
We have already paid the heavy cost to get them (the db/network hit), doing it this way is less lines of code and I think equivalent performance .

ericdill replied:
No, I mean don't you want to check the cache before you pay the db/network hit?

ericdill replied:
Though I suppose that is dangerous for the case of a run that is currently being collected, since event descriptors can (theoretically) be created at all times during the run and if you have event descriptors in the cache, you will not assume that you need to go check for more.

Though as I type this out, I suppose you could also check to see if there is a RunStop document for the given RunStart and, if so, you do not need to check the db for additional descriptors

tacaswell replied:
In this case we are finding event descriptors based on the _run start_ oid.  The caches are only keeping oid -> document mappings for single documents (`_*_CACHE[oid]['_id'] == oid`).  To check the local cache we would have to loop over the existing event descriptors and check the uid of their runstart against the uid passed in.  This is simpler ;)

ericdill commented on metadatastore/commands.py:
```diff
@@ -199,7 +427,7 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
+    uid : str
         Globally unique id string provided to metadatastore
     exit_status : {'success', 'abort', 'fail'}, optional
         indicating reason run stopped, 'success' by default
```
Should we be assuming success by default?

ericdill commented on metadatastore/commands.py:
```diff
@@ -214,56 +442,33 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
     run_stop : mongoengine.Document
```
This returns the uid

ericdill commented on metadatastore/commands.py:
```diff
@@ -292,30 +497,37 @@ def insert_event_descriptor(run_start, data_keys, time, uid=None,
         The document added to the collection.
```
This returns the uid

ericdill commented on metadatastore/commands.py:
```diff
@@ -276,11 +481,11 @@ def insert_event_descriptor(run_start, data_keys, time, uid=None,
     data_keys : dict
         Provides information about keys of the data dictionary in
-        an event will contain
+        an event will contain.  No key name may include '.'
```
Is this worth turning into the `.. warning:` syntax?

ericdill commented on metadatastore/commands.py:
```diff
@@ -336,7 +548,7 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
-    uid : str, optional
+    uid : str
         Globally unique id string provided to metadatastore
 
```
Add a Returns section for the uid

ericdill commented on metadatastore/commands.py:
```diff
@@ -345,34 +557,69 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
     # Allow caller to beg forgiveness rather than ask permission w.r.t
     # EventDescriptor creation.
     if descriptor is None:
         raise EventDescriptorIsNoneError()
```
Do we need this exception since descriptor a positional? Also the code that used this is dead in ophyd (and since it was my implementation, I can state with confidence that it was a bad design anyway)

tacaswell replied:
Gone.

tacaswell committed:
MNT: remove unneeded code

ericdill committed:
FIX: restore verbose __str__ code

tacaswell committed:
API: include '_' prefixed keys in `__contains__`

This means that not all elements that report `a in doc` are included
in iteration, but it does mean that for all emelement where `doc[a]`
does not raise, `(a in doc) == True`.

This breaks the object model less

tacaswell committed:
MNT: make vstr a bit for flexible

recursively call if `_name` in value, not by type filtering

tacaswell committed:
API: fix special casing for dataprotal Headers

tacaswell committed:
TST: push test coverage up to 100%

tacaswell committed:
PRF/API: apply what we learned about time ordering

As with `fetch_events_generator`, order events ascending by
time for performance boost.

tacaswell committed:
API: make Document readonly

With the caching scheme there are many shared references back to the
same underlying (mutable) object.  By making them read-only we will
prevent a whole class of impossible to track down bugs due to someone
doing a `pop` in user code and then the cache layer fallying over.

This seems like a better solution than making a copy on access.

tacaswell committed:
ENH: add helper function to get name, dict pair

tacaswell committed:
FIX: change Event not RunStart order

fixes and reverts f1cceb92dc75e2e250f74b5122f1eef4f54fad13

tacaswell committed:
API: raise custom error when no EventDescriptors

used to raise ValueError now raises NoEventDescriptors

tacaswell committed:
STY: whitespace

tacaswell committed:
ENH: helper to relpace a doc with it's uid

Get a new Document with a given reference field replaced by
it's uid.

tacaswell committed:
TST: push testing up to 100%

tacaswell committed:
PEP8: whitespace

tacaswell committed:
MNT: remove `__all__` as it is out-dated

tacaswell committed:
DOC: start to document

tacaswell committed:
API: eventdescriptor -> descriptor

Standardize all of the event descriptor related functions and
variables to 'descriptor' based names.

tacaswell committed:
API: find_event_desriptors -> find_descriptors

Added alias to maintain back-compatibility, deal with
deprecation later.

tacaswell committed:
MNT/API: standardize on runstart vs run_start

Existing functions are called in many of the tests/examples with
`func(run_start=bar, ...)` so punt on fixing those for now

tacaswell committed:
DOC: added docstrings + inline comments

tacaswell committed:
MNT/DOC: move and document _event_desc_given_oid

tacaswell committed:
MNT: _event_desc_given_oid->_descriptor_given_oid

For consistency

tacaswell committed:
API: event_desc_given_uid -> descriptor_given_uid

For consistency.  Not used outside of MDS yet so this is safe.

tacaswell committed:
DOC: document *_given_uid

tacaswell committed:
DOC/MNT: *_by_runstart

tacaswell committed:
API: eventdescriptors_by_runstart -> descriptors_by_runstart

Renamed for consistency

tacaswell committed:
MNT: simplify *_by_runstart

Do not actually use the full document so don't bother getting a copy of
it.

tacaswell committed:
DOC/MNT: fetch_events_generator

Document and clean up fetch_events_generator

tacaswell committed:
DOC: minor changes to fetch_events_table

tacaswell committed:
DOC: update and correct insert_run_start docs

tacaswell committed:
DOC/MNT: insert_run_stop

tacaswell committed:
DOC/MNT: insert_event_descriptor

ericdill commented on example/synthetic_cosine.py:
```diff
@@ -46,12 +45,14 @@
             'scalar_detector': [func(i) + np.random.randn() / 100,
                                 time.time()]}
-    e = insert_event(event_descriptor=e_desc, seq_num=idx,
+    e = insert_event(descriptor=e_desc, seq_num=idx,
```
rename `e_desc` local variable to `descriptor` or `desc`

ericdill commented on example/synthetic_cosine.py:
```diff
@@ -29,8 +27,9 @@
 
 custom = {'plotx': 'linear_motor', 'ploty': 'scalar_detector'}
```
drop plotx and ploty?

ericdill commented on example/synthetic_cosine.py:
```diff
@@ -1,14 +1,12 @@
-from metadatastore.api import (insert_run_start, insert_beamline_config,
-                               insert_event, insert_event_descriptor,
-                               insert_run_stop)
+from metadatastore.api import (insert_run_start,
```
this looks weird, move `insert_event` up to this line

tacaswell committed:
DOC: minor wording cahnge

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-                       Event, DataKey]:
-        collection._collection = None
+def _cache_runstart(rs):
+    """De-reference and cache a RunStart document
```
consider adding a `.. warning:` section since this mutates the rs dict

tacaswell replied:
dealt with differently

tacaswell committed:
API: insert_event_descriptor -> insert_descriptor

For consistency.

Left `insert_event_descriptor` in place as an alais, deprecate later

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-            'shape': shape_value},
-         'data_key2': {...}
-        }
+    rs : dict
```
`rs` is out of sync with the `runstop` input parameter

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+
+
+def _cache_runstop(runstop):
+    """De-reference and cache a RunStop document
```
consider adding a `.. warning:` section

ericdill replied:
because this function mutates the `runstop` input

tacaswell replied:
dealt with differently

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-    for collection in [RunStart, BeamlineConfig, RunStop, EventDescriptor,
-                       Event, DataKey]:
-        collection._collection = None
+def _cache_runstart(rs):
```
change this input parameter to `runstart` to be symmetric with the other `_cache_*` functions

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-    """Helper function for ophyd to format its data dictionary in whatever
-    flavor of the week metadatastore's spec says. This insulates ophyd from
-    changes to the mds spec
+def _cache_descriptor(descriptor):
```
consider adding a `.. warning:` section for input parameter mutation

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-    flavor of the week metadatastore's spec says. This insulates ophyd from
-    changes to the mds spec
+def _cache_descriptor(descriptor):
+    """De-reference and cache a RunStop document
```
`RunStop` -> `Descriptor`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-        {'data_key1': [data_value, timestamp_value],
-         'data_key2': [...],
-        }
+    descriptor : doc.Document
```
`metadatastore.doc.Document`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        Document instance for this EventDescriptor document.
+        The ObjectId has been stripped.
+    """
+    # pop the ObjectID
```
can probably drop this line comment

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+    return _cache_descriptor(descriptor)
+
+
+def runstop_by_runstart(runstart):
```
do you need the ensure connection decorator here?

tacaswell replied:
No, because the `runstart_given_uid` call (that I just added back) is decorated as such, but will add back just to be safe.

tacaswell committed:
MNT: make shallow copy of dicts before caching

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+    Returns
+    -------
+    descriptor : doc.Document
+        The RunStart document.
```
`RunStart` -> `EventDescriptor`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+    ------
+    NoRunStop
+        If no RunStop document exists for the given RunStart
+
```
remove this ~~whtie~~ white space

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        If no RunStop document exists for the given RunStart
+
+    """
+
```
don't need this white space either

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+    Parameters
+    ----------
+    runstart : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
```
`RunStop` -> `EventDescriptor`

tacaswell replied:
This one looks correct.

ericdill replied:
oh, yeah, you're right

tacaswell committed:
DOC/MNT: insert_event

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+    Parameters
+    ----------
+    descriptor : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
```
`The EventDescriptor` to get the `Event`s for

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+    """
+    descriptor_uid = doc_or_uid_to_uid(descriptor)
+    descriptor = descriptor_given_uid(descriptor_uid)
+    col = Event._get_collection()
```
why is this split from line 438?

ericdill commented on metadatastore/commands.py:
```diff
@@ -1,34 +1,94 @@
+# never drop these
+_RUNSTART_UID_to_OID_MAP = dict()
+_RUNSTOP_UID_to_OID_MAP = dict()
+_EVENTDESC_UID_to_OID_MAP = dict()
```
Consider renaming this to `_DESCRIPTOR_UID_TO_OID_MAP`

ericdill commented on metadatastore/commands.py:
```diff
@@ -1,34 +1,94 @@
+# implemented as.   Should move to uids asap
+_RUNSTART_CACHE_OID = boltons.cacheutils.LRU(max_size=1000)
+_RUNSTOP_CACHE_OID = boltons.cacheutils.LRU(max_size=1000)
+_EVENTDESC_CACHE_OID = boltons.cacheutils.LRU(max_size=1000)
```
Consider renaming this to `_DESCRIPTOR_CACHE_OID`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+
+    for ev in ev_cur:
+        # ditch the ObjectID
+        ev.pop('_id')
```
is `del` faster than `pop`?

tacaswell committed:
FIX: fix possible issue with caching

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        # pop the data
+        data = ev.pop('data')
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
```
`data.items()` -> `six.iteritems(data)`

ericdill replied:
You could also drop the line where you pop data off and just write this as `six.iteritems(ev['data'])`

tacaswell replied:
bah, got used to writing 3 only code.  I would rather leave the pop as I think it is a bit clearer.

ericdill replied:
Did some benchmarking with retrieval of 10k events and the two are no different in speed. I concur that you should leave it as `data = ev.pop('data')` and then iterate over data.items

ericdill replied:
but you should still change it to six.iteritems(data)

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        data = ev.pop('data')
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
+                                        for j in range(2)]
```
might want to just make `range(2)` explicit as `[1, 2]`

tacaswell replied:
well, `[0, 1]`

ericdill replied:
hah, right

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        EventDescriptor document
+
+    data_table : DataFrame
+        All of the data as a DataFrame indexed on sequence number
```
It is going to be challenging to combine events from multiple runs if these are indexed on sequence number because sequence numbers are degenerate across different run_start_uids.  Are sequence numbers also degenerate between the `event descriptors` of one RunStart?

As such, I would advocate for indexing these on event time, not sequence number.

ericdill replied:
this is no longer relevant as, after an in-person conversation with @danielballan and @tacaswell it was decided that this function should not use pandas at all and instead return lists and dicts of lists

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-    """Provide a head for a sequence of events. Entry point for an
-    experiment's run.
+def insert_run_start(time, scan_id, beamline_id, uid,
+                     owner='', group='', project='', custom=None):
```
Why did you change this from `None` to `''`

tacaswell replied:
Because the `None` was just being converted to `''` below.  Strings are immutable so they can be used as default arguments where as dicts are mutable and can not be.

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
     custom: dict, optional
         Any additional information that data acquisition code/user wants
-        to append to the Header at the start of the run.
+        to append to the RunStart at the start of the run.
```
Can you note that this `custom` dict is going to be unpacked into the top level of the RunStart document?

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
+        The timestamps of each of the measurements as a `DataFrame`.  Same
+        columns as `data_table` indexed on sequence number
+    """
+    import pandas as pd
```
Is it just getting fussy to do this?

``` python
try:
    pd
except NameError:
    import pandas as pd
    globals()['pd'] = pd
```

ericdill commented on metadatastore/commands.py:
```diff
@@ -211,76 +616,53 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
+    # create a reference field
+    rs_ref = DBRef('RunStart', runstart_oid)
+
+    run_stop = RunStop(run_start=rs_ref, reason=reason, time=time,
```
rename this local variable to `runstop`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,512 @@ def inner(*args, **kwargs):
-    experiment's run.
+def insert_run_stop(run_start, time, uid, exit_status='success',
+                    reason='', custom=None):
+    """Insert RunStop document into database
```
Is it worth noting that this is implemented so that only one `RunStop` can exist per `RunStart` and if a second one is created then it will raise?

tacaswell replied:
Noted.

ericdill commented on metadatastore/commands.py:
```diff
@@ -211,76 +616,53 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
+def insert_descriptor(run_start, data_keys, time, uid,
                             custom=None):
-    """ Create an event_descriptor in metadatastore database backend
+    """Inesrt an EventDescriptor document in to  database.
```
`Inesrt` -> `Insert`

ericdill replied:
remove the extra space between `to` and `database`

ericdill commented on metadatastore/commands.py:
```diff
@@ -211,76 +616,53 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
-@_ensure_connection
-def insert_event_descriptor(run_start, data_keys, time, uid=None,
+def insert_descriptor(run_start, data_keys, time, uid,
                             custom=None):
```
the white space on this line is offset from the `(`. It might be possible to stick `custom=None` on the previous line too?

ericdill commented on metadatastore/commands.py:
```diff
@@ -211,76 +616,53 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
-        if str:
-            uid of RunStart object to associate with this record
+    run_start : doc.Document or dict or str
+        The RunStart to get the RunStop for.  Can be either
```
`The RunStart that this `EventDescriptor` references. Can be either`

ericdill commented on metadatastore/commands.py:
```diff
@@ -211,76 +616,53 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
     data_keys : dict
         Provides information about keys of the data dictionary in
-        an event will contain
+        an event will contain.  No key name may include '.'
```
Where does one go to get the expected format of this `data_keys` dict?

tacaswell replied:
The ODM templates or the jsonschema files in bluesky.

ericdill replied:
Can you note that one should look at odm_templates.py or jsonschema in bluesky?

ericdill commented on metadatastore/commands.py:
```diff
@@ -288,34 +670,38 @@ def insert_event_descriptor(run_start, data_keys, time, uid=None,
+def insert_event(descriptor, time, seq_num, data, timestamps, uid):
     """Create an event in metadatastore database backend
 
     Parameters
```
The docstring for `descriptor` is wrong

ericdill commented on metadatastore/commands.py:
```diff
@@ -345,79 +731,78 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
+
+    events : iterable
+       iterable of dicts matching the bs.Event schema
 
```
Add `validate` to the parameters section

ericdill replied:
Add a `Returns` section

ericdill commented on metadatastore/commands.py:
```diff
@@ -345,79 +731,78 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
+    ----------
+    event_descriptor : str
+        The event descriptor uid that these events are associated with
+
```
drop the whitespace

tacaswell committed:
ENH: make IPython tab-complete work on docmuments

ericdill commented on metadatastore/commands.py:
```diff
@@ -336,88 +724,86 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
+            # transform the data to the storage format
+            val_ts_tuple = _transform_data(ev['data'], ev['timestamps'])
+            # create the mongoengine object, do this by hand if it goes
+            # faster
```
wait so is it a question whether it is faster to do it by hand?

tacaswell replied:
oh, already doing this by hand!

ericdill commented on metadatastore/commands.py:
```diff
@@ -604,48 +953,18 @@ def find_run_starts(**kwargs):
+    return (_cache_runstart(rs) for rs in rs_objects.order_by('-time'))
 
 
 @_ensure_connection
```
The docstring for the run_start parameter needs to be changed to metadatastore.doc.Document

tacaswell replied:
done I think?  This seems to have come un-morred from it's line.

ericdill replied:
yeah, this is done

ericdill commented on metadatastore/commands.py:
```diff
@@ -790,25 +1106,34 @@ def find_events(descriptor=None, **kwargs):
-            for ev in events)
+    for ev in events:
+        # ditch the ObjectID
+        ev.pop('_id')
```
`pop` -> `del`

ericdill commented on metadatastore/commands.py:
```diff
@@ -790,25 +1106,34 @@ def find_events(descriptor=None, **kwargs):
+        # pop the data
+        data = ev.pop('data')
+        # replace it with the friendly paired dicts
+        ev['data'], ev['timestamps'] = [{k: v[j] for k, v in data.items()}
```
`for k, v in six.iteritems(ev['data'].items())}`

ericdill replied:
and remove line 1129

ericdill replied:
err remove line `data = ev.pop('data')`

tacaswell committed:
MNT: first round of @ericdill edits

ericdill commented on metadatastore/commands.py:
```diff
@@ -790,25 +1106,34 @@ def find_events(descriptor=None, **kwargs):
+        # pop the descriptor oid
+        desc_oid = ev.pop('descriptor_id')
+        # replace it with the defererenced descriptor
+        ev['descriptor'] = _descriptor_given_oid(desc_oid)
```
could probably make this `ev['descriptor'] = _descriptor_given_oid(ev.pop('descriptor_id'))` and delete the line where you capture `desc_oid`

ericdill commented on metadatastore/doc.py:
```diff
@@ -0,0 +1,224 @@
+        ret : dict
+            Data payload of Document
+
+
```
remove the white space

ericdill commented on metadatastore/odm_templates.py:
```diff
@@ -10,29 +10,13 @@
 
 
-__all__ = ['BeamlineConfig', 'RunStart', 'RunStop', 'DataKey',
+__all__ = ['RunStart', 'RunStop', 'DataKey',
```
You can fit more on this line

ericdill commented on metadatastore/commands.py:
```diff
@@ -790,25 +1116,34 @@ def find_events(descriptor=None, **kwargs):
     if 'event_descriptor_id' in kwargs:
         raise ValueError("Use 'descriptor_id', not 'event_descriptor_id'.")
```
This error message is out of date

ericdill commented on metadatastore/doc.py:
```diff
@@ -0,0 +1,224 @@
+from __future__ import (absolute_import, division, print_function,
+                        unicode_literals)
```
I thought we weren't using `unicode_literals` any longer

ericdill commented on metadatastore/doc.py:
```diff
@@ -0,0 +1,224 @@
+  </tr>
+{% endfor %}
+</table>
+"""
```
Is this the same repr that was in the old document.Document? If not, @danielballan are you still :+1: on this html table repr?  I remember at some point recently you were not happy with it.

ericdill replied:
Err, if it is the same table repr

danielballan replied:
My feeling was more that notebooks aren't a great context for data exploration. I don't have a better idea for an HTML repr for this, so let's leave it.

ericdill replied:
ok

ericdill commented on metadatastore/doc.py:
```diff
@@ -0,0 +1,224 @@
+
+    Parameters
+    ----------_
+    name : str, optional
```
`name` -> `doc`

ericdill commented on metadatastore/test/test_commands.py:
```diff
@@ -316,12 +332,141 @@ def test_normalize_human_friendly_time():
+    # Create an EventDescriptor that indicates the data
+    # keys and serves as header for set of Event(s)
+    mdsc.insert_descriptor(data_keys=data_keys,
+                                 time=ttime.time(),
```
text is tabbed in too far

ericdill commented on example/synthetic_cosine.py:
```diff
@@ -27,14 +24,15 @@
 # keys and serves as header for set of Event(s)
-e_desc = insert_event_descriptor(data_keys=data_keys, time=time.time(),
+descriptor = insert_descriptor(data_keys=data_keys, time=time.time(),
                                  run_start=rs)
```
Need a `uid` field here and the whitespace is off

ericdill commented on example/synthetic_cosine.py:
```diff
@@ -46,15 +44,17 @@
                      time=time.time(),
-                     data=data)
+                     data=data,
+                     uid=str(uuid.uuid4()))
```
this needs a `timestamp` field

ericdill commented on example/synthetic_cosine.py:
```diff
@@ -46,15 +44,17 @@
             'Tsam': [i + 5, time.time()],
             'scalar_detector': [func(i) + np.random.randn() / 100,
                                 time.time()]}
```
`data` needs to be split into `data` and `timestamps`

tacaswell committed:
API: don't expose ODM templates, they are private

tacaswell committed:
STY: clean up whitespace in api.py

tacaswell committed:
API/PRF: return base types from fetch_events_table

No longer have optoinal pandas dep

tacaswell committed:
MNT: local variable rename

ericdill committed:
PRF: faster method to convert data format

tacaswell committed:
DOC: docstring edits

tacaswell committed:
DOC: fix vstr docstring

tacaswell committed:
TST: simplify tests / whitespace

tacaswell committed:
MNT/DOC: yet more clean up

tacaswell committed:
MNT: short-circuit if unique key in input

tacaswell committed:
MNT: clean up local variables

tacaswell committed:
PRF: more performance tuning

faster in python3, same performance in python2

tacaswell committed:
FIX: bug in super order

tacaswell committed:
MNT/STY: minor changes

tacaswell committed:
DOC: YMDC

tacaswell committed:
FIX: fix over-zealous simplification

tacaswell committed:
API: break all the APIs because consistency

run_start -> runstart
run_stop -> runstop

almost everywhere practical.

Document field are unchanged so the input `run_start` on the insert_*
functions must remain unchanged or bluesky is severely broken.

tacaswell committed:
MNT: simplify ODM indices

tacaswell committed:
MNT: fix demo

tacaswell committed:
MNT: removed `__all__` from odm_templates

tacaswell committed:
STY: whitespace

tacaswell committed:
DOC: cleaned up odm docs

tacaswell committed:
PRF: use faster iteration

tacaswell committed:
STY: sundry whitespace cleanup

ericdill commented on metadatastore/commands.py:
```diff
@@ -336,88 +747,85 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
+        a Document/dict with a 'uid' key or a uid string
+    events : iterable
+       iterable of dicts matching the bs.Event schema
 
```
Add a `Returns` section

ericdill replied:
Add `validate` to the `Parameters` section

ericdill commented on metadatastore/commands.py:
```diff
@@ -1,34 +1,94 @@
-           'format_data_keys', 'format_events', 'reorganize_event']
+# process local caches of 'header' documents these are storing object indexed
+# on ObjectId because that is what the reference fields in mongo are
+# implemented as.   Should move to uids asap
```
Just some minor grammar stuff

```
# process local caches of 'header' documents. These caches are storing objects indexed
# on ObjectId because that is how the reference fields are implemented in mongo
# Should move to uids asap
```

ericdill commented on metadatastore/commands.py:
```diff
@@ -1,34 +1,94 @@
+
+       This has no effect if the connection already exists, simply changing
+       the values in `conf.connection_config` is not sufficient to change what
+       data base is being used.
```
Do you need to also call `db_disconnect()` for this function to auto-connect to a different database after the config info is changed in `conf.connection_config`?

[edit] clarifying the question

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
-    the event descriptor
+    Returns
+    -------
+    runstart : doc.Document
```
This should probably say `runstop` not `runstart`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
-        {'data_key1':
-         'data_key2': mds.odm_templates.DataKeys
-        }
+    runstart : doc.Document
```
This should probably say `descriptor` or `eventdescriptor`, not `runstart`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+    NoRunStop
+        If no RunStop document exists for the given RunStart
+    """
+
```
remove this blank line

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+        {'run_start_id': oid})
+
+    if runstop is None:
+        raise NoRunStop("No run stop exists")
```
``` python
        raise NoRunStop("No run stop exists for a RunStart with uid=%s" % runstart_uid)
```

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+
+    # if nothing found, raise
+    if not rets:
+        raise NoEventDescriptors("No EventDescriptors exists")
```
``` python
raise NoRunStop("No EventDescriptors exist for a RunStart with uid=%s" % runstart_uid)
```

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+    Parameters
+    ----------
+    descriptor : doc.Document or dict or str
+        The EventDescriptors to get the Events for.  Can be either
```
Remove the plural from `EventDescriptors`. `descriptor` is not a list

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+    for ev in ev_cur:
+        # ditch the ObjectID
+        del ev['_id']
+        # pop the descriptor oid
```
`# delete the descriptor oid`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+
+
+def _transpose(in_data, keys, field):
+    out = {k: [None] * len(in_data) for k in keys}
```
Would it be possible to get a docstring on this?

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+
+    uids : list
+        The uid of each event.
+
```
can you remove all the extra whitespace from this Returns section

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+    ----------
+    descriptor : dict or str
+        The EventDestriptor to get the events for.  Can be either
+        a dict or a uid.
```
Pattern matching on the other dict or str inputs

```
    descriptor : doc.Document or dict or str
        A Document/dict with a 'uid' key or a uid string
```

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+    data_table : dict
+        dict of lists of the transposed data
+
+    uids : list
```
But delete this one, not the one that I left the comment on, sorry

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+    return descriptor, data_table, seq_nums, times, uids, timestamps_table
+
+
+def db_disconnect():
```
could you move these two functions up to where `_ensure_connection` is defined? They seem rather out of place in the file here

ericdill replied:
by these two i mean `db_disconnect` and `db_connect`

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
+
+    keys = list(descriptor['data_keys'])
+
+    # TODO make sure this is fast
```
_transpose is definitely fast enough...

https://gist.github.com/a7dae88858ca7e3eec14

ericdill commented on metadatastore/commands.py:
```diff
@@ -39,167 +99,524 @@ def inner(*args, **kwargs):
-            The metadatastore RunStart document
-        if str:
-            uid of RunStart object to associate with this record
+    runstart : doc.Document or dict or str
```
out of sync with the input parameter `run_start`

ericdill commented on metadatastore/commands.py:
```diff
@@ -207,124 +624,118 @@ def insert_run_stop(run_start, time, uid=None, exit_status='success',
-            The metadatastore RunStart document
-        if str:
-            uid of RunStart object to associate with this record
+    runstart : doc.Document or dict or str
```
out of sync with the `run_start` input parameter

ericdill commented on metadatastore/commands.py:
```diff
@@ -336,88 +747,85 @@ def insert_event(descriptor, time, seq_num, data, timestamps=None, uid=None):
+
+    Parameters
+    ----------
+    descriptor : doc.Document or dict or str
```
out of sync with the `event_descriptor` input parameter

ericdill commented on metadatastore/commands.py:
```diff
@@ -516,49 +924,15 @@ def _normalize_human_friendly_time(val):
     Parameters
     ----------
+    runstart : doc.Document or dict or str
+        The RunStart to search event for.  Can be either
```
I dont understand why `event` is in here

ericdill commented on metadatastore/commands.py:
```diff
@@ -516,49 +924,15 @@ def _normalize_human_friendly_time(val):
 
     Parameters
     ----------
+    runstart : doc.Document or dict or str
```
It is a little unclear why we are accepting doc.Document as a valid option here. If you already have the runstart document, why are you searching for it again? This should just be `str`

ericdill replied:
Actually we should drop this kwarg entirely and just accept `uid`

ericdill commented on metadatastore/commands.py:
```diff
@@ -675,42 +1018,36 @@ def find_run_stops(run_start=None, **kwargs):
+            try:
+                yield runstop_by_runstart(runstart)
+            except NoRunStop:
+                return
```
hmmm, I'm not sure I like this pattern of returning None instead of raising a NoRunStop

tacaswell committed:
FIX: cache might be out of date

ericdill commented on metadatastore/doc.py:
```diff
@@ -0,0 +1,221 @@
+    """Recursive document walker and formatter
+
+    Parameters
+    ----------_
```
delete this trailing `_`

ericdill commented on metadatastore/api.py:
```diff
@@ -1,16 +1,6 @@
-                       find_beamline_configs, find_run_starts, find_last,
-                       find_run_stops)
+from .commands import (find_events, find_descriptors, find_runstarts,
+                       find_last, find_runstops)
```
Can you add the `runstop_by_runstart` and similar functions?

tacaswell committed:
DOC: doc changes

tacaswell committed:
MNT: move functions around

tacaswell committed:
MNT: don't serach for RunStart with a RunStart

As you seem to already have it

tacaswell committed:
FIX: fix find_runstops

tacaswell committed:
DOC: fixup

tacaswell committed:
FIX: properly guard against None

tacaswell committed:
API: simplify names of

runstop_by_runstart -> stop_by_start
descriptors_by_runstart -> descriptors_by_start

tacaswell committed:
API: revert earlier API changes

insert_runstart backto insert_run_start

tacaswell committed:
API: revert earlier changes

insert_runstop backto insert_run_stop
find_runstop backto find_run_stop

tacaswell committed:
STY: remove whitespace

tacaswell committed:
DOC: add API changes entry

tacaswell committed:
DOC: add schema reference
